File: Chapter-7-Rapid-Development-Practices/browser-automation-mvn-selenium/src/main/java/SeleniumTest1.java
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.safari.SafariDriver;
import org.openqa.selenium.htmlunit.HtmlUnitDriver;  //Headless: java based implementation. For any other language binding Selenium Server required to use this driver.
import org.openqa.selenium.support.ui.ExpectedCondition;
import org.openqa.selenium.support.ui.WebDriverWait;

public class SeleniumTest1  {
    public static void main(String[] args) {
        
       // WebDriver driver = new FirefoxDriver();
//		  WebDriver driver = new SafariDriver();
		WebDriver driver = new HtmlUnitDriver();  //HtmlUnitDriver(true) blows up though

		
// Additional work is required for Chrome...		
//  http://code.google.com/p/selenium/wiki/ChromeDriver
//  import org.openqa.selenium.chrome.ChromeDriver;
//  WebDriver driver = new ChromeDriver(); 

        driver.get("http://www.google.com");  //or driver.navigate().to("http://www.google.com");
        WebElement element = driver.findElement(By.name("q"));
        element.sendKeys("Cheese!");
		element.submit();  //  WebDriver finds the form from the element

        // Check the title of the page
        System.out.println("Page title is: " + driver.getTitle());
        
        // Google's search is rendered dynamically with JavaScript.
        // Wait for the page to load, timeout after 10 seconds
        (new WebDriverWait(driver, 10)).until(new ExpectedCondition<Boolean>() {
            public Boolean apply(WebDriver d) {
                return d.getTitle().toLowerCase().startsWith("cheese!");
            }
        });

        System.out.println("Page title is: " + driver.getTitle());
        driver.quit();
    }
}

File: Chapter-7-Rapid-Development-Practices/browser-automation-mvn-selenium/src/main/scripts/seleniumtest1.rb
require 'rubygems'
require 'selenium-webdriver'

#driver = Selenium::WebDriver.for :firefox
driver = Selenium::WebDriver.for :safari
driver.get "http://google.com"

element = driver.find_element :name => "q"
element.send_keys "Cheese!"

#element.clear
#element.send_keys "JavaScript"

# http://docs.seleniumhq.org/docs/03_webdriver.jsp#selenium-webdriver-api-commands-and-operations
#driver.find_element(:partial_link_text, 'Mozilla').click
#driver.find_element(:id, "x")

element.submit

puts "Page title is #{driver.title}"

wait = Selenium::WebDriver::Wait.new(:timeout => 10)
wait.until { driver.title.downcase.start_with? "cheese!" }

puts "Page title is #{driver.title}"

driver.quit